### DISCUSSION TITLE
How to determine Result<A>.Match(Func...) return type?

### DISCUSSION THREAD

**Discussion by logany-hi** [louthy/language-ext#1052] [General]
The below code errors saying the type arguments for the method  `Result<Member>.Match<R>(Func<Member, R>, Func<Exception, R>)` can't be inferred from usage.

Obviously I'm doing something wrong but based on the video about this from Nick Chapsas, I really can't see WHAT I'm doing wrokg. I don't need this method to even return a type...
```
public async Task<IActionResult> OnGetAsync()
{
    var memberResult = await memberService.GetMemberFromPrincipal(User); // This returns Result<Member>.
    var _member = memberResult .Match(b =>
    {
        var result = MapMemberForView(b);
        this.Member = result;

        return new OkResult();
    },
    exception =>
    {
        this.Error = exception;
        return new BadRequestResult();
    });
}
```

In fact, what would suit me best would be something like this:
```
public async Task OnGetAsync()
{
    var memberResult = await memberService.GetMemberFromPrincipal(User); // This returns Result<Member>.
    memberResult.Match(member =>
    {
        this.Member = MapMemberForView(member);
    },
    exception =>
    {
        this.Error = exception;
    });
}
```

Why can't I do this?

Nick has his method as...

```
var result = await _customerService.CreateAsync(customer);

return result.Match(b =>
{
    var customerResponse = customer.ToCustomerResponse();
    return CreatedAtAction("Get", new { customerResponse.Id }, customerResponse...; // the line cuts off here
}, exception =>
{
    if (exception is ValidationException validationException)
    {
        return BadRequest(validationException.ToProblemDetails());
    }
}
```

Doesn't seem any different to my code; success and fail return different types but mine doesn't work...

**Comment by WarrenCrabb** [comment]
I'm pretty sure it is because OkResult and BadRequestResult can not be implicitly converted to the same type. Try changing your `var _member` to `IActionResult _member`.

It works in your later example because it is immediately being returned so the return type can be inferred from the function return type which I assume is ` Task<IActionResult>`.

**Comment by logany-hi** [comment]
I don't NEED a return type here though; does that mean this isn't appropriate use in this case?

This seems to work (no compile time errors)... I don't know well enough to verify if it's ideal or not.

The code includes all the work I'm doing at this level so it's a more complete example.

```
public async Task OnGetAsync()
{
    var memberResult = await memberService.GetMemberFromPrincipal(User);
    await memberResult.Match(async _member =>
    {
        this.Member = MapMemberForView(_member);
        await SetAssessmentsAsync(_member.Id);
        await SetUpcomingAssessmentAsync(_member.Id);
        return _member;
    },
    exception =>
    {
        this.Error = exception;
        return null;
    });
}
```

**Comment by WarrenCrabb** [comment]
Match has an overload that takes an [Action](https://louthy.github.io/language-ext/LanguageExt.Core/Monads/Alternative%20Value%20Monads/Option/Option/index.html#Prelude_0_match_1 ). Probably more appropriate for your last example since you don't appear to do anything with the return.
